be/src/exec/pipeline/exchange/exchange_parallel_merge_source_operator.cpp

```diff
@@ -0,0 +1,121 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "exec/pipeline/exchange/exchange_parallel_merge_source_operator.h"
+
+#include "exec/sort_exec_exprs.h"
+#include "runtime/data_stream_mgr.h"
+#include "runtime/data_stream_recvr.h"
+#include "runtime/exec_env.h"
+
+namespace starrocks::pipeline {
+
+Status ExchangeParallelMergeSourceOperator::prepare(RuntimeState* state) {
+    RETURN_IF_ERROR(SourceOperator::prepare(state));
+    auto* factory = down_cast<ExchangeParallelMergeSourceOperatorFactory*>(_factory);
+    _stream_recvr = factory->get_stream_recvr(state, _unique_metrics);
+    _merger = factory->get_merge_path_merger(state);
+    _merger->bind_profile(_driver_sequence, _unique_metrics.get());
+    return Status::OK();
+}
+
+bool ExchangeParallelMergeSourceOperator::has_output() const {
+    if (_is_finished) {
+        return false;
+    }
+    if (_merger->is_current_stage_finished(_driver_sequence)) {
+        return false;
+    }
+    if (_merger->is_pending(_driver_sequence)) {
+        return false;
+    }
+    return true;
+}
+
+bool ExchangeParallelMergeSourceOperator::is_finished() const {
+    if (_limit < 0) {
+        return _is_finished;
+    } else {
+        return _num_rows_returned >= _limit || _is_finished;
+    }
+}
+
+Status ExchangeParallelMergeSourceOperator::set_finishing(RuntimeState* state) {
+    _is_finished = true;
+    auto* factory = down_cast<ExchangeParallelMergeSourceOperatorFactory*>(_factory);
+    factory->close_stream_recvr();
+    return Status::OK();
+}
+
+StatusOr<ChunkPtr> ExchangeParallelMergeSourceOperator::pull_chunk(RuntimeState* state) {
+    // TODO process limit
+    auto chunk = _merger->try_get_next(_driver_sequence);
```

Comment:
is it possible that try_get_next returns a nullptr?

Reply:
Yes, all the merge process are driven by this method. In most stage, it will return nullptr.